packages/ai/core/util/chat-store.ts

```diff
@@ -0,0 +1,602 @@
+  ToolInvocationUIPart,
+  UIMessage,
+} from '../types';
+
+interface ChatStoreSubscriber {
+  onChatMessagesChanged?: (id: string) => void;
+  onChatStatusChanged?: (id: string) => void;
+  onChatErrorChanged?: (id: string) => void;
+  // onChatAdded?
+  // onChatRemoved?
+}
+
+export enum ChatStoreEvent {
+  ChatMessagesChanged = 'chat-messages-changed',
+  ChatStatusChanged = 'chat-status-changed',
+  ChatErrorChanged = 'chat-error-changed',
+  // ChatAdded = 'chat-added',
+  // ChatRemoved = 'chat-removed',
+}
+
+const ChatStoreEventMap = {
+  [ChatStoreEvent.ChatMessagesChanged]: 'onChatMessagesChanged',
+  [ChatStoreEvent.ChatStatusChanged]: 'onChatStatusChanged',
+  [ChatStoreEvent.ChatErrorChanged]: 'onChatErrorChanged',
+} as const;
+
+export interface ChatStoreInitialization {
+  chats?: Record<string, Pick<ChatState, 'messages'>>;
+  onChatStoreChange?: ({
+    event,
+    chatId,
+    state,
+  }: {
+    event: ChatStoreEvent;
+    chatId: string;
+    state: ChatState;
+  }) => void;
+}
+
+export interface ChatState {
+  status: 'submitted' | 'streaming' | 'ready' | 'error';
+  messages: UIMessage[];
+  step?: number;
+  activeResponse?: {
+    message: UIMessage;
+    partialState: {
+      textPart?: TextUIPart;
+      reasoningPart?: ReasoningUIPart;
+      reasoningTextDetail?: { type: 'text'; text: string; signature?: string };
+      tools?: Record<
+        string,
+        { step: number; text: string; index: number; toolName: string }
+      >;
+    };
+  };
+  error?: Error;
+}
+
+export class ChatStore {
+  private chats: Map<string, ChatState>;
+  private subscribers: Set<ChatStoreSubscriber>;
+  private onChatStoreChange?: ChatStoreInitialization['onChatStoreChange'];
+
+  constructor({ chats = {}, onChatStoreChange }: ChatStoreInitialization = {}) {
+    this.chats = new Map(
+      Object.entries(chats).map(([id, state]) => [
+        id,
+        {
+          messages: [...state.messages],
+          status: 'ready',
+          activeResponse: undefined,
+          error: undefined,
+          step: 0,
+        },
+      ]),
+    );
+    this.subscribers = new Set();
+    this.onChatStoreChange = onChatStoreChange;
+  }
+
+  get totalChats() {
+    return this.chats.size;
+  }
+
+  getStatus(id: string) {
+    const chat = this.chats.get(id);
+    if (!chat) return;
+    return chat.status;
+  }
+
+  setStatus({ id, status }: { id: string; status: ChatState['status'] }) {
+    const chat = this.chats.get(id);
+    if (!chat || chat.status === status) return;
+    chat.status = status;
+    this.emitEvent({ id, event: ChatStoreEvent.ChatStatusChanged });
+  }
+
+  getError(id: string) {
+    const chat = this.chats.get(id);
+    if (!chat) return;
+    return chat.error;
+  }
+
+  setError({ id, error }: { id: string; error: ChatState['error'] }) {
+    const chat = this.chats.get(id);
+    if (!chat) return;
+    chat.error = error;
+    this.emitEvent({ id, event: ChatStoreEvent.ChatErrorChanged });
+  }
```

Comment:
are these indirections required or can the `Chat` object itself be exposed?

Reply:
(might require the introduction of that class)

Reply:
this might potentially be an even better abstraction - what if we had a "Chat" object which could be passed into `useChat`? what are the pros/cons vs passing ChatStore + id?

Reply:
oooooo i'd like to ensure chatstore is the central manager, introducing some sort of chat class does sound interesting and would be a lovely abstraction but am initially concerned about de-centralizing state management - let me play around and get back to you with a stronger opinion!

Reply:
sounds good!